File: backend/services/youtube_service.py
"""YouTube transcript extraction service."""

from typing import Optional

from pydantic import BaseModel
from youtube_transcript_api import YouTubeTranscriptApi
from youtube_transcript_api._errors import (
    NoTranscriptFound,
    TranscriptsDisabled,
    VideoUnavailable,
)

from database.models import Transcript, TranscriptSegment
from database.repositories.transcript_repository import TranscriptRepository
from utils.youtube_utils import build_youtube_url, extract_video_id


class TranscriptResult(BaseModel):
    """Result of a transcript extraction attempt."""

    success: bool
    video_id: str
    transcript: Optional[Transcript] = None
    error_type: Optional[str] = None
    error_message: Optional[str] = None


class YouTubeService:
    """Service for extracting and managing YouTube transcripts."""

    def __init__(self):
        self.transcript_repo = TranscriptRepository()
        self.api = YouTubeTranscriptApi()

    def get_transcript(
        self,
        video_id_or_url: str,
        languages: list[str] = ["en", "hi", "es", "fr", "de", "pt", "ja", "ko", "zh"],
        use_cache: bool = True,
    ) -> TranscriptResult:
        """
        Get the transcript for a YouTube video.

        Args:
            video_id_or_url: YouTube video ID or URL
            languages: Preferred languages in order
            use_cache: Whether to use cached transcripts

        Returns:
            TranscriptResult with success status and transcript or error info
        """
        video_id = extract_video_id(video_id_or_url)
        if not video_id:
            return TranscriptResult(
                success=False,
                video_id=video_id_or_url,
                error_type="InvalidURL",
                error_message="Could not extract a valid YouTube video ID from the provided URL.",
            )

        # Check cache first
        if use_cache:
            cached = self.transcript_repo.get_by_video_id(video_id)
            if cached:
                return TranscriptResult(
                    success=True,
                    video_id=video_id,
                    transcript=cached,
                )

        # Fetch from YouTube using the new API
        try:
            # First, list available transcripts to find the best one
            transcript_list = self.api.list(video_id)

            # Try to find a transcript in preferred languages
            transcript_data = None
            is_generated = False
            language_code = "en"

            # First try manually created transcripts
            try:
                transcript = transcript_list.find_manually_created_transcript(languages)
                transcript_data = transcript.fetch()
                is_generated = False
                language_code = transcript.language_code
            except NoTranscriptFound:
                # Fall back to auto-generated
                try:
                    transcript = transcript_list.find_generated_transcript(languages)
                    transcript_data = transcript.fetch()
                    is_generated = True
                    language_code = transcript.language_code
                except NoTranscriptFound:
                    # Try any available transcript
                    available = list(transcript_list)
                    if available:
                        transcript = available[0]
                        transcript_data = transcript.fetch()
                        is_generated = transcript.is_generated
                        language_code = transcript.language_code

            if not transcript_data:
                return TranscriptResult(
                    success=False,
                    video_id=video_id,
                    error_type="NoTranscriptFound",
                    error_message="No transcript found for this video in any language.",
                )

            # Convert to our model - handle both dict and FetchedTranscriptSnippet
            segments = []
            for seg in transcript_data:
                # The new API returns objects with attributes, not dicts
                if hasattr(seg, 'text'):
                    segments.append(TranscriptSegment(
                        text=seg.text,
                        start=seg.start,
                        duration=seg.duration,
                    ))
                else:
                    # Fallback for dict format
                    segments.append(TranscriptSegment(
                        text=seg["text"],
                        start=seg["start"],
                        duration=seg["duration"],
                    ))

            transcript_obj = Transcript(
                video_id=video_id,
                video_url=build_youtube_url(video_id),
                language_code=language_code,
                is_generated=is_generated,
                segments=segments,
            )

            # Cache the transcript
            try:
                self.transcript_repo.create(transcript_obj)
            except Exception:
                # Ignore cache errors, we still have the transcript
                pass

            return TranscriptResult(
                success=True,
                video_id=video_id,
                transcript=transcript_obj,
            )

        except TranscriptsDisabled:
            return TranscriptResult(
                success=False,
                video_id=video_id,
                error_type="TranscriptsDisabled",
                error_message="Transcripts are disabled for this video by the owner.",
            )
        except VideoUnavailable:
            return TranscriptResult(
                success=False,
                video_id=video_id,
                error_type="VideoUnavailable",
                error_message="This video is unavailable, private, or has been removed.",
            )
        except Exception as e:
            error_type = type(e).__name__
            return TranscriptResult(
                success=False,
                video_id=video_id,
                error_type=error_type,
                error_message=str(e),
            )

    def get_available_languages(self, video_id_or_url: str) -> dict:
        """
        Get available transcript languages for a video.

        Args:
            video_id_or_url: YouTube video ID or URL

        Returns:
            Dict with 'manual' and 'generated' language lists
        """
        video_id = extract_video_id(video_id_or_url)
        if not video_id:
            return {"error": "Invalid video ID", "manual": [], "generated": []}

        try:
            transcript_list = self.api.list(video_id)

            manual = []
            generated = []

            for transcript in transcript_list:
                lang_info = {
                    "code": transcript.language_code,
                    "name": transcript.language,
                }
                if transcript.is_generated:
                    generated.append(lang_info)
                else:
                    manual.append(lang_info)

            return {"manual": manual, "generated": generated}

        except Exception as e:
            return {"error": str(e), "manual": [], "generated": []}

    def get_cached_transcript(self, video_id: str) -> Optional[Transcript]:
        """Get a cached transcript if it exists."""
        return self.transcript_repo.get_by_video_id(video_id)

    def clear_cache(self, video_id: str) -> bool:
        """Clear the cached transcript for a video."""
        return self.transcript_repo.delete_by_video_id(video_id)


def get_transcript_error_help(error_type: str) -> str:
    """
    Get user-friendly help text for transcript errors.

    Args:
        error_type: The error type from TranscriptResult

    Returns:
        Helpful message for the user
    """
    error_help = {
        "TranscriptsDisabled": """
**Transcripts are disabled for this video.**

The video owner has disabled transcripts/captions for this video.

**What you can try:**
1. Check if the video has community-contributed captions
2. Use YouTube's auto-generated captions directly on YouTube
3. Try a different video that has captions enabled
""",
        "VideoUnavailable": """
**This video is unavailable.**

The video may be private, age-restricted, or has been removed.

**What you can try:**
1. Verify the URL is correct
2. Check if you can access the video directly on YouTube
3. Make sure the video is publicly available
""",
        "NoTranscriptFound": """
**No transcript available for this video.**

This video doesn't have transcripts in any supported language.

**What you can try:**
1. Check if the video has auto-generated captions on YouTube
2. Use YouTube's "Show transcript" feature (click ... below the video)
3. Try a different video with captions
""",
        "InvalidURL": """
**Invalid YouTube URL.**

The URL provided doesn't appear to be a valid YouTube video URL.

**Supported formats:**
- youtube.com/watch?v=VIDEO_ID
- youtu.be/VIDEO_ID
- youtube.com/embed/VIDEO_ID
""",
    }

    return error_help.get(
        error_type,
        f"""
**Unable to extract transcript.**

An error occurred while trying to get the transcript: {error_type}

**What you can try:**
1. Check your internet connection
2. Try again in a few moments
3. Use YouTube's built-in transcript feature
""",
    )


# Global YouTube service instance
youtube_service = YouTubeService()

File: src/services/youtube-service.ts
/**
 * YouTube service for interacting with the backend YouTube API.
 *
 * This service provides methods for:
 * - Extracting YouTube transcripts
 * - Managing cached transcripts
 * - Getting available languages
 */

import { API_BASE_URL } from "@/config/app-config";

export interface TranscriptSegment {
  text: string;
  start: number;
  duration: number;
}

export interface Transcript {
  id: string;
  video_id: string;
  video_url: string;
  title?: string;
  language_code: string;
  is_generated: boolean;
  segments: TranscriptSegment[];
  full_text: string;
  created_at: string;
}

export interface TranscriptResponse {
  success: boolean;
  video_id?: string;
  transcript?: Transcript;
  error_type?: string;
  error_message?: string;
}

export interface LanguageInfo {
  code: string;
  name: string;
  is_generated: boolean;
  is_translatable: boolean;
}

class YouTubeService {
  private baseUrl: string;

  constructor() {
    this.baseUrl = `${API_BASE_URL}/api/v1/youtube`;
  }

  /**
   * Extract transcript from a YouTube URL or video ID.
   */
  async getTranscript(
    urlOrVideoId: string,
    languageCode?: string
  ): Promise<TranscriptResponse> {
    const response = await fetch(`${this.baseUrl}/transcript`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        url: urlOrVideoId,
        language_code: languageCode,
      }),
    });

    if (!response.ok) {
      const error = await response.json().catch(() => ({}));
      return {
        success: false,
        error_type: "request_failed",
        error_message: error.detail || response.statusText,
      };
    }

    return response.json();
  }

  /**
   * Get a cached transcript by video ID.
   */
  async getCachedTranscript(videoId: string): Promise<TranscriptResponse> {
    const response = await fetch(`${this.baseUrl}/transcript/${videoId}`, {
      method: "GET",
      headers: {
        "Content-Type": "application/json",
      },
    });

    if (!response.ok) {
      if (response.status === 404) {
        return {
          success: false,
          video_id: videoId,
          error_type: "not_found",
          error_message: "Transcript not found in cache",
        };
      }
      return {
        success: false,
        video_id: videoId,
        error_type: "request_failed",
        error_message: response.statusText,
      };
    }

    return response.json();
  }

  /**
   * Delete a cached transcript.
   */
  async clearCache(videoId: string): Promise<boolean> {
    const response = await fetch(`${this.baseUrl}/transcript/${videoId}`, {
      method: "DELETE",
      headers: {
        "Content-Type": "application/json",
      },
    });

    return response.ok;
  }

  /**
   * Get available languages for a video.
   */
  async getAvailableLanguages(videoId: string): Promise<LanguageInfo[]> {
    const response = await fetch(`${this.baseUrl}/languages/${videoId}`, {
      method: "GET",
      headers: {
        "Content-Type": "application/json",
      },
    });

    if (!response.ok) {
      return [];
    }

    const data = await response.json();
    return data.languages || [];
  }

  /**
   * Extract video ID from a YouTube URL.
   */
  extractVideoId(url: string): string | null {
    const patterns = [
      /(?:youtube\.com\/watch\?v=|youtu\.be\/|youtube\.com\/embed\/|youtube\.com\/v\/|youtube\.com\/shorts\/)([a-zA-Z0-9_-]{11})/,
      /^([a-zA-Z0-9_-]{11})$/,
    ];

    for (const pattern of patterns) {
      const match = url.match(pattern);
      if (match) {
        return match[1];
      }
    }

    return null;
  }

  /**
   * Check if a string contains a YouTube URL.
   */
  containsYouTubeUrl(text: string): boolean {
    const pattern =
      /(?:youtube\.com\/watch\?v=|youtu\.be\/|youtube\.com\/embed\/|youtube\.com\/v\/|youtube\.com\/shorts\/)[a-zA-Z0-9_-]{11}/;
    return pattern.test(text);
  }

  /**
   * Find all YouTube URLs in a text.
   */
  findYouTubeUrls(text: string): string[] {
    const pattern =
      /(?:https?:\/\/)?(?:www\.)?(?:youtube\.com\/watch\?v=|youtu\.be\/|youtube\.com\/embed\/|youtube\.com\/v\/|youtube\.com\/shorts\/)([a-zA-Z0-9_-]{11})/g;
    const matches = text.matchAll(pattern);
    return Array.from(matches, (m) => m[0]);
  }
}

// Export singleton instance
export const youtubeService = new YouTubeService();
export default youtubeService;
